firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/datagrams.txt

```diff
@@ -0,0 +1,446 @@
+
+    WHEN("A body with a complete 2-byte header, and paylaod buffer as '0x00' consistent with the length field of the header is parsed")
+      THEN("The parse method reports ok status")
+      THEN("After the parse method is called, The value returned by the seq accessor method is equal to the sequence field of the input_buffer body's header")
+      THEN("The value returned by the length accessor method is equal to the length field of the input_buffer body's header")
+      THEN("The payload returned from the payload accessor method is equal to the payload from the input_buffer body")
+      THEN("the payload accessor method returns a reference to the payload buffer given in the Datagram constructor") 
+      THEN("the payload given in the Datagram constructor is independent of the input buffer body")
+
+    WHEN("A body with a complete 2-byte header, and a non-empty payload buffer where the length field of the header is inconsistent with the actual length of the payload buffer")
+      THEN("The parse method reports ok status")
+      THEN("After the parse method is called, The value returned by the seq accessor method is equal to the sequence field of the input_buffer body's header")
+      THEN("The value returned by the length accessor method is equal to the length field of the input_buffer body's header")
+      THEN("The value returned by the length accessor method is not equal to the size of the payload buffer in the input_buffer body")
+      THEN("The payload returned from the payload accessor method is equal to the payload from the input_buffer body")
+      THEN("the payload accessor method returns a reference to the payload buffer given in the Datagram constructor") 
+      THEN("the payload given in the Datagram constructor is independent of the input buffer body")
+
+    WHEN("A body with sequence field of the header inconsistent with sequence given in the constructor of the datagram")
+      THEN("The parse method reports ok status")
+      THEN("After the parse method is called, The value returned by the seq accessor method is equal to the sequence field of the input_buffer body's header")
+      THEN("The value returned by the length accessor method is equal to the length field of the input_buffer body's header")
+      THEN("The payload returned from the payload accessor method is equal to the payload from the input_buffer body")
+      THEN("the payload accessor method returns a reference to the payload buffer given in the Datagram constructor") 
+      THEN("the payload given in the Datagram constructor is independent of the input buffer body")
+
+
+Scenario: Datagram correctly preserves data in roundtrip writing and parsing
+  GIVEN("A datagram constructed with a non-empty payload buffer and sequence number equal to 0")
+    WHEN("A body with seq, length and payload is generated, parsed into a new ParsedDatagram object, and generated again with a ConstructedDatagram object")
```

Comment:
In this test, each section has a THEN which expects a different payload. I think the payload is supposed to be the same on each one?

Reply:
I kept it consistent with CRCElement where we used different payload buffers for input to Parsed and Constructed CRCElements, do you want me to change it ?

Reply:
I don't mean using different payload buffers (we should do that), I mean that after the first write you assert that the payload is written into the body as `0x61 0x6a 0x1a 0x6a 0x29 0xcf 0x01 0x81 0xbe 0x9d`, then after the parse you assert that the payload is parsed from the body as `0x00`, and then after the second write you assert that the payload is written into the body as `0x56 0xd6 0x42 0xc5 0xe1 0xf0 0x30 0xe5 0xc8 0x4d`

Reply:
> then after the parse you assert that the payload is parsed from the body as 0x00

If this is the only concern then that was a mistake and I fixed it.

> then after the second write you assert that the payload is written into the body as 0x56 0xd6 0x42 0xc5 0xe1 0xf0 0x30 0xe5 0xc8 0x4d

Is this a concern as well? Here I constructed a new ConstructedDatagram object passing the payload as `0x56 0xd6 0x42 0xc5 0xe1 0xf0 0x30 0xe5 0xc8 0x4d` and called write after that.

Reply:
The point of the roundtrip is that we're transforming data (generating a body), then passing the result of that (the body) into the inverse (parsing a body), then passing the result of that (the payload) into the inverse of the inverse (generating a body) - so we should be using the output of the previous step as the output of the next step.

Reply:
Pseudocode as discussed in meeting:
```
payload_input -> first write's ConstructedDatagram constructor
first write's ConstructedDatagram write -> first_body_output
payload_output -> ParsedDatagram constructor
first_body_output -> ParseDatagram parse
(check payload_output)
payload_output -> second write's ConstructedDatagram constructor
second write's ConstructedDatagram write -> second_body_output
```

Reply:
> Pseudocode as discussed in meeting:
> 
> ```
> payload_input -> first write's ConstructedDatagram constructor
> first write's ConstructedDatagram write -> first_body_output
> payload_output -> ParsedDatagram constructor
> first_body_output -> ParseDatagram parse
> (check payload_output)
> payload_output -> second write's ConstructedDatagram constructor
> second write's ConstructedDatagram write -> second_body_output
> ```

I think you had mentioned in the meeting to keep the `ParsedDatagram constructor` payload empty.
other than that I have fixed and pushed.

Here is the code:
https://github.com/pez-globo/pufferfish-software/blob/feature/firmware-test-protocols-datagrams/firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/Datagrams.cpp#L736